### PR TITLE
Removed reference counting of GraphicBuffers.

### PR DESCRIPTION
**Pull request by emilwestergren**
It's not needed for our internal buffers.

@sebastianbaginski peer plz

### CODE REVIEW

**Comment by sebastianbaginski** [comment]
You don't need reference counting but still you can benefit from automatic memory management thanks to the smart pointers. I always use them even for simplest objects, they cost almost nothing. Are you sure you want to revert to plain pointers ?
Although I'm not familiar with this `sp` class, you can always do
```C++
typedef std::shared_ptr<GraphicsBuffer> GraphicsBufferPtr;
```
and use `GraphicsBufferPtr` like a normal pointer. Even later just change typedef to `GraphicsBuffer*` if you think shared pointers are causing your code to lag.

**Comment by emilwestergren** [comment]
The other option is to allocate the smart pointers on the heap instead. The reason is because I pass ownership to Vidhance, so I can't let all smart pointers in the wrapper be destroyed and free the buffer.

Pass ownership to Vidhance:
```c++
sp<GraphicBuffer>* heapPointer = new sp<GraphicBuffer>();
*heapPointer = new GraphicBuffer();
kean_graphicBuffer_new(heapPointer, ...);
```

Pass ownership back to wrapper and free:
```c++
sp<GraphicBuffer>* heapPointer = (sp<GraphicBuffer>*)handle;
delete(heapPointer);
```

Would you prefer that?

**Comment by sebastianbaginski** [comment]
No, in that case definitely not, I didn't know about passing the ownership to external library.

File: HAL/CameraWrapper.cpp [hunk 1]
```diff
@@ -14,14 +14,25 @@ CameraWrapper::CameraWrapper() {
...
 }
-sp<GraphicBuffer> CameraWrapper::getCameraBuffer(int width, int height, int format, int usage) {
+void CameraWrapper::freeCameraBuffers() {
 	pthread_mutex_lock(&this->cameraBufferMutex);
-	sp<GraphicBuffer> result = NULL;
-	std::vector< sp<GraphicBuffer> >::iterator iter;
+	std::vector<GraphicBuffer*>::iterator iter;
 	for(iter = this->cameraBuffers.begin(); iter != this->cameraBuffers.end(); ++iter) {
 		sp<GraphicBuffer> buffer = *iter;
+		buffer = NULL;
```
Comment (sebastianbaginski) [inline HAL/CameraWrapper.cpp]: Can you compile this code with `-std=c++11` ? This way you could use the `auto` keyword (and lots of other nice features).
```C++
for (auto * buffer : this->cameraBuffers) delete buffer;
```

File: HAL/CameraWrapper.cpp [hunk 2]
```diff
@@ -14,14 +14,25 @@ CameraWrapper::CameraWrapper() {
...
+	}
+	this->cameraBuffers.clear();
+	pthread_mutex_unlock(&this->cameraBufferMutex);
+}
+GraphicBuffer* CameraWrapper::getCameraBuffer(int width, int height, int format, int usage) {
+	pthread_mutex_lock(&this->cameraBufferMutex);
+	GraphicBuffer* result = NULL;
+	std::vector<GraphicBuffer*>::iterator iter;
+	for(iter = this->cameraBuffers.begin(); iter != this->cameraBuffers.end(); ++iter) {
+		GraphicBuffer* buffer = *iter;
```
Comment (sebastianbaginski) [inline HAL/CameraWrapper.cpp]: You can just rename `iter` to `buffer` and it should work, iterators have `operator->`. The syntax could also benefit from the `auto` loops from current C++ standard.

**Comment by sebastianbaginski** [comment]
Alright, there is some room for syntax cleanups in general. If you are sure that you got the memory management right after reverting from `sp` then I'm done.

File: HAL/CameraWrapper.cpp (hunk 1 above)
Comment (sebastianbaginski) [inline HAL/CameraWrapper.cpp]: Or if you delete & clear containers a lot then I find this
```C++
template<typename Container>
void deleteAll(Container& container) {
    for (auto * it : container) delete it;
    container.clear();
}
```
quite useful.

Comment (emilwestergren) [inline HAL/CameraWrapper.cpp]: nice

File: HAL/CameraWrapper.cpp (hunk 2 above)
Comment (emilwestergren) [inline HAL/CameraWrapper.cpp]: Still need to dereference once. Iterator->GraphicBuffer*->GraphicBuffer

**Pull request merged**